### ISSUE TITLE
Bluetooth audio stutter

### ISSUE THREAD

**Issue by mastix** [CyanogenMod/android_kernel_samsung_klte#3]
Hey guys,

I hope that this is the correct place for a bug report.

When using my BT headphones, I'm experiencing heavy audio stuttering. That does not happen on other (non-CM) devices.

Device: Samsung Galaxy S5 (klte)
CM: 12.1 Nightly (17th of May)

Is there anything I need to do? Is this a known issue?

Regards,

Sascha

**Comment by T3charmy** [comment]
A temp workaround till this issue is fixed, is to turn off wifi, and change the setting to only allow wifi scanning when wifi is on.

**Comment by Whoopie** [comment]
Did you try adding "ro.bluetooth.request.master=true" in the /system/build.prop file?

**Comment by JensErat** [comment]
Do I understand correctly you have to add this line to build configuration, and compile the image from scratch? Seems I have to dig out my CM build environment once again...

**Comment by Whoopie** [comment]
No, you just need to edit your /system/build.prop by adding the mentioned line in your ROM.

**Comment by JensErat** [comment]
I hope I applied the settings correctly:

    adb root
    adb remount
    adb shell
    # echo "ro.bluetooth.request.master=true" >>/system/build.prop
     # reboot

After reboot:

    adb shell
    $ grep bluetooth /system/build.prop
    ro.bluetooth.request.master=true

Sadly, it does not seem to have resolved the issue, still having multiple seconds long stutters. They might have got rarer, though; but I'm not sure about this. At least I had only one larger stutter while listening 15 minutes to music (and several short cracks in the beginning), usually I'd have more of them. I'll let the change applied for a while though to get a better glance on improvements (and will report again in some days).

**Comment by Rythyrix** [comment]
Leaving build.prop editing out, it appears on my G900FD running nightly 20151125 that this issue is confined to 2.4ghz networks, connecting to a family member's 5ghz / 2.4ghz network in 5ghz mode does not stutter. Swapped between the two to make sure. Can anyone confirm?

**Comment by JensErat** [comment]
I cannot completely confirm connecting 5GHz wireless resolves the issue, although there might be some improvement (I'm yet to update to the newest nightlies, though).

I can confirm there is _some_ interference between WiFi and Bluetooth; disabling WiFi completely removes most of the stutters, using the [Mozilla Location Stumbler](https://location.services.mozilla.com/) (a service scanning for mobile and WiFi access points) heavily increases issues (as the phone is doing permanent scans for WiFi networks).

**Comment by mastix** [comment]
Have you guys been able to fix this? I'm thinking about going back to CM. But since Bluetooth Audio is very important for me, this would be a reason not to install CM. So sad...

**Comment by JensErat** [comment]
I don't think anybody even got a _clue_ right now what's wrong here.

**Comment by qwertxzy** [comment]
I still get this, why didn't anyone fix it in 8 months

**Comment by FernandoMiguel** [comment]
getting this on my Nexus 6p, with latest nightly.
turning off works around it, but makes the device pretty limited.
no issue with stock 6.0.1

**Comment by FernandoMiguel** [comment]
disabling  "Bluetooth HCI snoop log" seem to improve for me.
but i'm currently connected to a 5GHz network, so i need to test on 2.4GHz to be sure

**Comment by JensErat** [comment]
@FernandoMiguel are you sure about _disabling_? On my device it wasn't active at all while I have those problems.

**Comment by FernandoMiguel** [comment]
@JensErat mine was set to Check, and not disabled.
there are 3 options.
and as I said, i need to test on 2.4GHz, since that where it conflits with BT freq. at home i have 5GHz, so it's harder to say.
but on my test, i had no stutter

**Comment by Rythyrix** [comment]
@ family members home right now. On 5gh, no stutter. 13.0 nightly 20160122 for klteduos. When using 2.4ghz, have stutter when traffic is used on WiFi.

**Comment by JensErat** [comment]
Following [an article on the Galaxy S5's wireless details](http://s4gru.com/index.php?/blog/1/entry-363-teaser-samsung-galaxy-s5-gets-a-boost-via-wi-fi-but-not-carrier-aggregation/), it seems the Galaxy S5 is one of the first WiFi MIMO capable phones (with other words, it is able to transmit two streams at the same time featuring two antennas). Looking at the diagram, antenna 1 is for both WiFi and BT, while antenna 2 is only doing WiFi.

The Nexus 6p also features MIMO. Is anybody aware of similar problems with other phones, and do those also feature MIMO?

This would not provide any solution, but at least hint that there's some connection between features and BT disturbances. And might enable somebody familiar with the wireless stacks to dig into what's going wrong here. The following is wild speculation: there seems to be something like an antenna broker if I look at several chip diagrams of the S5. Maybe something allows both to use the same antenna the same time, which will result in transmission errors, while the coding managing antenna access only properly checks for different frequency bands, not different users of the same band. Or has a weird scheduling of the antennas.

**Comment by FernandoMiguel** [comment]
just to note that i had zero issues with 6.x stock
only began having terrible audio gaps on CM.
Both Moto Nexus 6  and my very very old Moto G 2013 dual sim also experience audio gaps with CM12.1/13 and my Motorokr S305 headset
i hightly dought the MotoG has MIMO :) but it does have very little RAM, and that tends to kill the apps or make them slow down a lot.

**Comment by cepardov** [comment]
Device: klte (Samsung Galaxy S5)
Firmware: CM13 Stock Kernel
Inssue: BT Malfunction Stream Audio
Fixed: https://github.com/CyanogenMod/android_kernel_samsung_klte/commit/951e2ff60348cc47839cf8dc5aa8ffa775622431
Firmware: CM 12.1 Stock Kernel
State: Partialy Fixed

**Comment by JensErat** [comment]
Did you quote from some other bug report system?

This patch is from summer last year. I guess most here will be running notably newer snapshots already having the patch, but it does not seem this resolved the issue.

**Comment by cepardov** [comment]
Stuttering diminishes , but does not eradicate it completely (Sorry for my english)

**Comment by FernandoMiguel** [comment]
i did some more testing with my Nexus 6p. 
wondering around the street, not connected to any WiFi (even any in range, London underground), it would shutter 
:(
